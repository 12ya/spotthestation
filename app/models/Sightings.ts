import { types } from "mobx-state-tree"

export const Sighting = types.model("Sighting", {
  date: types.optional(types.string, ""),
  maxHeight: types.maybeNull(types.number),
  minAzimuth: types.maybeNull(types.number),
  maxAzimuth: types.maybeNull(types.number),
  minAltitude: types.maybeNull(types.number),
  maxAltitude: types.maybeNull(types.number),
  visible: types.maybeNull(types.number),
  notify: types.maybeNull(types.boolean),
  dayStage: types.maybeNull(types.number),
})
