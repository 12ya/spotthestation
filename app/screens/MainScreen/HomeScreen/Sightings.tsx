import React, { useEffect, useState } from "react"
import { ViewStyle, View, PressableProps, TextStyle, ScrollView } from "react-native"
import Modal from "react-native-modal"
import { Button, Icon, Text, IconTypes, Toggle } from "../../../components"
import { colors, fontSizes, lineHeights, scale, typography } from "../../../theme"
import { ExpandContainer } from "../components/ExpandContainer"
import { ListItem } from "../components/ListItem"
import { useSafeAreaInsetsStyle } from "../../../utils/useSafeAreaInsetsStyle"
import { addDays } from "date-fns"
import { formatDateWithTZ, getShortTZ } from "../../../utils/formatDate"
import { ISSSighting } from "../../../services/api"
import { getCalendars } from 'expo-localization'
import * as storage from "../../../utils/storage"
import { normalizeHeight } from "../../../utils/normalizeHeight"

export interface SightingsProps {
  sightings: ISSSighting[]
  isUS?: boolean
  isNotifyAll?: boolean
  timezone?: string
  onClose?: PressableProps["onPress"]
  onToggle?: (values: ISSSighting) => void
  onToggleAll?: (value: boolean) => void
}

export function Sightings({ onClose, sightings, onToggle, onToggleAll, isUS, isNotifyAll, timezone }: SightingsProps) {
  const $marginTop = useSafeAreaInsetsStyle(["top"], "margin")
  const $paddingBottom = useSafeAreaInsetsStyle(["bottom"], "padding")
  const [sightingsCoachVisible, setSightingsCoachVisible] = useState(false)

  const formatedDate = (date: string): string => {
    const timeFormat = getCalendars()[0].uses24hourClock ? "H:mm" : "h:mm A"
    const shortTZ = getShortTZ(timezone)
    if (formatDateWithTZ(date, `YYYY-MM-DD`, { timeZone: timezone }) === formatDateWithTZ(new Date().toISOString(), `YYYY-MM-DD`, { timeZone: timezone })) return `Today, ${formatDateWithTZ(date, timeFormat, { timeZone: timezone })} ${shortTZ}`
    if (formatDateWithTZ(date, `YYYY-MM-DD`, { timeZone: timezone }) === formatDateWithTZ(addDays(new Date(), 1).toISOString(), `YYYY-MM-DD`, { timeZone: timezone })) return `Tomorrow, ${formatDateWithTZ(date, timeFormat, { timeZone: timezone })} ${shortTZ}`
    return `${formatDateWithTZ(date, `${isUS ? "MMM DD, YYYY" : "DD MMM YYYY"}, ${timeFormat}`, { timeZone: timezone })} ${shortTZ}`
  }

  const setStageIcon = (stage): { icon: IconTypes, color: string} => {
    switch(stage) {
      case 0: return { icon: 'moon', color: colors.palette.neutral450 }
      case 1: return { icon: 'sunset', color: colors.palette.nasaOrange }
      case 2: return { icon: 'sun', color: colors.palette.yellow }
      default: return { icon: 'sunset', color: colors.palette.nasaOrange }
    }
  }

  const getCoach = async () => {
    setSightingsCoachVisible(!(await storage.load('sightingsCoachVisible')))
  }

  useEffect(() => {
    getCoach().catch((e) => console.log(e))
  }, [])

  const handleSetSightingsCoachVisible = async () => {
    setSightingsCoachVisible(false)
    await storage.save('sightingsCoachVisible', true)
  }

  return (
    <View style={[$modalBodyContainer, $marginTop, $paddingBottom]}>
      <Icon icon="x" 
        accessible
        accessibilityLabel="x button"
        accessibilityHint="close modal"
        accessibilityRole="button"
        color={colors.palette.neutral450} 
        onPress={onClose} 
        containerStyle={$close} 
        size={36}
      />
      <Text
        accessible
        accessibilityLabel="title"
        accessibilityHint="title"
        accessibilityRole="text"
        tx="homeScreen.selectSightings.title"
        style={$title}
      />
      <Text
        accessible
        accessibilityLabel="title"
        accessibilityHint="title"
        accessibilityRole="text"
        tx="homeScreen.selectSightings.selectMessage"
        style={$selectMessageText}
      />
      <View style={$switchContainer}>
        <Text tx="homeScreen.selectSightings.switch" style={$label} />
        <Toggle
          accessible
          accessibilityLabel="switch button"
          accessibilityHint="toggle notifications"
          variant="switch" 
          value={isNotifyAll} 
          onValueChange={() => {
            onToggleAll(!isNotifyAll)
          }}
        />
      </View>
      <View style={$scrollContainer}>
        <ExpandContainer title="homeScreen.selectSightings.sightings" expandble={false} reverseTitle>
          <ScrollView
            accessible
            accessibilityLabel="Sightings scrollable area"
            accessibilityHint="Sightings scrollable area"
            accessibilityRole="scrollbar"
          >
            {[...sightings].filter(item => new Date(item.date) > new Date()).map((sighting: ISSSighting) =>
              <ListItem
                key={sighting.date}
                icon="clock"
                secondIcon={setStageIcon(sighting.dayStage)}
                title={formatedDate(sighting.date)}
                selected={sighting.notify}
                subtitle={`Above the horizon ${sighting.visible} min`}
                withSwitch
                onToggle={() => onToggle(sighting)}
              />
            )}
          </ScrollView>
        </ExpandContainer>
      </View>
      {sightingsCoachVisible && <Modal
        isVisible={sightingsCoachVisible}
        useNativeDriver
        useNativeDriverForBackdrop
        backdropOpacity={.4}
        style={$modal}
      >
        <View 
          accessible
          accessibilityLabel="coach mark"
          accessibilityHint="coach mark"
          accessibilityRole="text"
          style={[$coachModalBodyContainer, { marginTop: normalizeHeight(.2) }]}
        >
          <View style={$legend}>
            <Icon icon="sun" size={44} color={colors.palette.yellow} />
            <Text tx="homeScreen.selectSightings.coach.sun" style={$body} />
          </View>
          <View style={$legend}>
            <Icon icon="sunset" size={44} color={colors.palette.nasaOrange} />
            <Text tx="homeScreen.selectSightings.coach.sunset" style={$body} />
          </View>
          <View style={$legend}>
            <Icon icon="moon" size={44} color={colors.palette.neutral450} />
            <Text tx="homeScreen.selectSightings.coach.moon" style={$body} />
          </View>
          <Button
            accessible
            accessibilityLabel="dismiss button"
            accessibilityHint="dismiss coach mark"
            tx="homeScreen.coachMarks.dismiss"
            textStyle={$nextButtonText}
            style={$nextButton}
            pressedStyle={$nextButton}
            onPress={handleSetSightingsCoachVisible}
          />
        </View>
      </Modal>}
    </View>
  )
}

const $modalBodyContainer: ViewStyle = {
  backgroundColor: colors.palette.neutral350,
  borderTopLeftRadius: scale(18),
  borderTopRightRadius: scale(18),
  flex: 1
}

const $coachModalBodyContainer: ViewStyle = {
  backgroundColor: colors.palette.buttonBlue,
  borderRadius: scale(16),
  paddingVertical: 36,
  paddingHorizontal: 30,
  width: '100%',
}

const $scrollContainer: ViewStyle = {
  paddingHorizontal: scale(36),
  flex: 1,
  paddingBottom: scale(30)
}

const $close: ViewStyle = {
  position: "absolute",
  top: 0,
  right: 0,
  padding: scale(18),
  zIndex: 5
}

const $title: TextStyle = {
  marginTop: scale(10),
  marginBottom: scale(10),
  fontFamily: typography.primary?.normal,
  fontSize: fontSizes[28],
  lineHeight: lineHeights[44],
  color: colors.palette.neutral250,
  paddingHorizontal: scale(36),
}

const $selectMessageText: TextStyle = {
  width: "95%",
  fontFamily: typography.primary?.normal,
  fontSize: fontSizes[18],
  lineHeight: lineHeights[22],
  color: colors.palette.neutral100,
  paddingHorizontal: scale(36),
}

const $switchContainer: ViewStyle = {
  flexDirection: "row",
  justifyContent: "space-between",
  borderBottomColor: colors.palette.neutral350,
  paddingHorizontal: scale(36),
  paddingTop: scale(15),
}

const $label: TextStyle = {
  color: colors.palette.neutral250,
  fontSize: fontSizes[16],
  fontFamily: typography.primary.normal,
  lineHeight: lineHeights[21],
  width: '80%'
}

const $modal: ViewStyle = {
  flex: 1,
  left: 0,
  margin: 0,
  paddingHorizontal: 18,
  justifyContent: 'flex-start',
}

const $legend: ViewStyle = {
  flexDirection: "row",
  margin: 0,
  justifyContent: "space-between",
  width: "80%"
}

const $body: TextStyle = {
  fontFamily: typography.primary.normal,
  fontSize: fontSizes[18],
  lineHeight: lineHeights[22],
  color: colors.palette.neutral100,
  paddingBottom: 10,
  paddingLeft: 5
}

const $nextButton: ViewStyle = {
  height: scale(56),
  backgroundColor: colors.palette.neutral100,
  borderRadius: scale(28),
  borderWidth: 0,
  width: scale(140),
  alignSelf: 'center',
  marginTop: 24
}

const $nextButtonText: TextStyle = {
  fontFamily: typography.primary.medium,
  fontSize: fontSizes[18],
  lineHeight: lineHeights[22],
  color: colors.palette.buttonBlue,
}
