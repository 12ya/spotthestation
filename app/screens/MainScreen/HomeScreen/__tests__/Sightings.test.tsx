/* eslint-disable react/display-name */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import React from "react"
import { Sightings } from "../Sightings"
import { act, render } from "@testing-library/react-native"
import MockDate from "mockdate"

describe("Sightings", () => {
  beforeEach(() => {
    MockDate.set("12-12-2012 10:10:10")
  })

  afterEach(() => {
    MockDate.reset()
  })

  it("renders correctly", async () => {
    const component = render(
      <Sightings
        timeOfDay=""
        duration=""
        onDurationChange={jest.fn()}
        onTimeOfDayChange={jest.fn()}
        sightings={[
          {
            date: "2020-12-12T10:10:10",
            maxHeight: 20,
            visible: 5,
            dayStage: 0,
            notify: true,
            minAzimuth: 0,
            maxAzimuth: 0,
            minAltitude: 0,
            maxAltitude: 0,
          },
          {
            date: "2020-12-13T10:15:10",
            maxHeight: 20,
            minAzimuth: 0,
            maxAzimuth: 0,
            minAltitude: 0,
            maxAltitude: 0,
            visible: 5,
            dayStage: 1,
            notify: false,
          },
          {
            date: "2020-12-24T10:20:10",
            maxHeight: 20,
            minAzimuth: 0,
            maxAzimuth: 0,
            minAltitude: 0,
            maxAltitude: 0,
            visible: 5,
            dayStage: 2,
            notify: true,
          },
        ]}
      />,
    )

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 100))
    })

    expect(component.toJSON()).toMatchSnapshot()
  })
})
