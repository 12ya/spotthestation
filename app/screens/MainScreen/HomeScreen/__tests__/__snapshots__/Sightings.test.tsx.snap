// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#1E2022",
        "borderTopLeftRadius": 26,
        "borderTopRightRadius": 26,
        "flex": 1,
      },
      Object {},
      Object {},
    ]
  }
>
  <View
    accessibilityHint="close modal"
    accessibilityLabel="x button"
    accessibilityRole="button"
    accessible={true}
    style={
      Object {
        "padding": 26,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 5,
      }
    }
  >
    <RCTImageView
      internal_analyticTag={null}
      resizeMode="contain"
      source={
        Array [
          Object {
            "height": undefined,
            "uri": undefined,
            "width": undefined,
          },
        ]
      }
      style={
        Object {
          "height": 52,
          "overflow": "hidden",
          "resizeMode": "contain",
          "tintColor": "#ADADAE",
          "width": 52,
        }
      }
      tintColor="#ADADAE"
    />
  </View>
  <Text
    accessibilityHint="title"
    accessibilityLabel="title"
    accessibilityRole="text"
    accessible={true}
    allowFontScaling={false}
    style={
      Array [
        Object {},
        Array [
          Object {
            "fontSize": 23,
            "lineHeight": 35,
          },
          Object {
            "fontFamily": "Helvetica Neue",
          },
          Object {
            "color": "#191015",
          },
        ],
        undefined,
        undefined,
        Object {
          "color": "#DEDDDE",
          "fontFamily": "Helvetica Neue",
          "fontSize": 40,
          "lineHeight": 63,
          "marginBottom": 15,
          "marginTop": 15,
          "paddingHorizontal": 52,
        },
      ]
    }
  >
    homeScreen.selectSightings.title undefined
  </Text>
  <Text
    accessibilityHint="title"
    accessibilityLabel="title"
    accessibilityRole="text"
    accessible={true}
    allowFontScaling={false}
    style={
      Array [
        Object {},
        Array [
          Object {
            "fontSize": 23,
            "lineHeight": 35,
          },
          Object {
            "fontFamily": "Helvetica Neue",
          },
          Object {
            "color": "#191015",
          },
        ],
        undefined,
        undefined,
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Helvetica Neue",
          "fontSize": 26,
          "lineHeight": 32,
          "paddingHorizontal": 52,
          "width": "95%",
        },
      ]
    }
  >
    homeScreen.selectSightings.selectMessage undefined
  </Text>
  <View
    style={
      Object {
        "borderBottomColor": "#1E2022",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 52,
        "paddingTop": 22,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {},
          Array [
            Object {
              "fontSize": 23,
              "lineHeight": 35,
            },
            Object {
              "fontFamily": "Helvetica Neue",
            },
            Object {
              "color": "#191015",
            },
          ],
          undefined,
          undefined,
          Object {
            "color": "#DEDDDE",
            "fontFamily": "Helvetica Neue",
            "fontSize": 23,
            "lineHeight": 30,
            "width": "80%",
          },
        ]
      }
    >
      homeScreen.selectSightings.switch undefined
    </Text>
    <View
      accessibilityHint="toggle notifications"
      accessibilityLabel="switch button"
      accessibilityRole="switch"
      accessibilityState={
        Object {
          "checked": undefined,
          "disabled": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Array [
                Object {
                  "alignItems": "center",
                  "borderWidth": 3,
                  "flexDirection": "row",
                  "flexGrow": 0,
                  "flexShrink": 0,
                  "height": 35,
                  "justifyContent": "space-between",
                  "overflow": "hidden",
                  "width": 35,
                },
                Object {
                  "borderRadius": 23,
                  "borderWidth": 0,
                  "height": 40,
                  "width": 66,
                },
              ],
              Object {
                "backgroundColor": "#353739",
              },
              undefined,
            ]
          }
        >
          <View
            animatedStyle={
              Object {
                "value": Object {
                  "opacity": 0,
                },
              }
            }
            collapsable={false}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#0B3D91",
                "borderColor": "rgba(0, 0, 0, 0)",
                "height": "100%",
                "opacity": 0,
                "overflow": "hidden",
                "paddingEnd": 6,
                "paddingStart": 6,
                "position": "absolute",
                "width": "100%",
              }
            }
          />
          <View
            animatedStyle={
              Object {
                "value": Object {
                  "marginStart": 6,
                  "start": "0%",
                },
              }
            }
            collapsable={false}
            style={
              Object {
                "backgroundColor": "#F4F2F1",
                "borderRadius": 17,
                "height": 29,
                "marginStart": 6,
                "position": "absolute",
                "start": "0%",
                "width": 29,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "paddingBottom": 43,
        "paddingHorizontal": 52,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "marginTop": 52,
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 26,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "flexDirection": "row-reverse",
              },
            ]
          }
        >
          <Text
            accessibilityHint="homeScreen.selectSightings.sightings"
            accessibilityLabel="title"
            accessibilityRole="text"
            accessible={true}
            allowFontScaling={false}
            style={
              Array [
                Object {},
                Array [
                  Object {
                    "fontSize": 23,
                    "lineHeight": 35,
                  },
                  Object {
                    "fontFamily": "Helvetica Neue",
                  },
                  Object {
                    "color": "#191015",
                  },
                ],
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "#ADADAE",
                    "fontFamily": "Helvetica Neue",
                    "fontSize": 19,
                    "lineHeight": 23,
                    "textTransform": "uppercase",
                  },
                  Object {
                    "textAlign": "right",
                    "width": "100%",
                  },
                ],
              ]
            }
          >
            homeScreen.selectSightings.sightings undefined
          </Text>
        </View>
      </View>
      <RCTScrollView
        accessibilityHint="Sightings scrollable area"
        accessibilityLabel="Sightings scrollable area"
        accessibilityRole="scrollbar"
        accessible={true}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</View>
`;
